### ISSUE TITLE
CNN accuracy drops if trained on multiple GPUs

### ISSUE THREAD

**Issue by pi-null-mezon** [davisking/dlib#1681]
I have found strange (in terms of prediction accuracy) issue when trained simple network on two GPUs. The learning problem was also very simple (I have feeded the network by full balck and full white images at trainng time, and asked it to minimize loss_multiclass_log for this two classes). Actually I am working on different task, but task is not important for the issue, so I remove all that is not really matter.

## Expected Behavior
If I build my example code with ```dnn_trainer<net_type> trainer(net,sgd())``` and it run on a machine with two GPUs:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.78       Driver Version: 410.78       CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 2070    Off  | 00000000:09:00.0 Off |                  N/A |
| 15%   41C    P0     1W / 175W |      0MiB /  7949MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce RTX 2070    Off  | 00000000:0A:00.0 Off |                  N/A |
| 27%   52C    P2   169W / 175W |   1587MiB /  7952MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    1     20346      C   Learner                                     1575MiB |
+-----------------------------------------------------------------------------+
``` 
I got excelent (100 % validation accuracy) result:
```
trainobjs.size(): 2
validobjs.size(): 2
step#: 0     learning rate: 0.1   average loss: 0            steps without apparent progress: 0
step#: 830   learning rate: 0.1   train loss: 0.00126927   test loss: 0.000182292  steps without apparent progress: train=804, test=161
step#: 1680  learning rate: 0.1   train loss: 0.00011663   test loss: 0.000140182  steps without apparent progress: train=1662, test=334
step#: 2530  learning rate: 0.01  train loss: 0.000124164  test loss: 0.000131655  steps without apparent progress: train=1917, test=505
step#: 3380  learning rate: 0.001  train loss: 0.000123289  test loss: 0.000127507  steps without apparent progress: train=1067, test=675
step#: 4230  learning rate: 0.001  train loss: 0.000123396  test loss: 0.000125391  steps without apparent progress: train=1998, test=845
step#: 5080  learning rate: 1e-05  train loss: 0.000123314  test loss: 0.000124505  steps without apparent progress: train=1392, test=815
step#: 5930  learning rate: 1e-05  train loss: 0.000123028  test loss: 0.000123763  steps without apparent progress: train=918, test=985
Training has been accomplished
Saved state to ./trainer_11_sync
Accuracy evaluation on validation set:
iter #0 - accuracy: 1
iter #1 - accuracy: 1
iter #2 - accuracy: 1
iter #3 - accuracy: 1
iter #4 - accuracy: 1
iter #5 - accuracy: 1
iter #6 - accuracy: 1
iter #7 - accuracy: 1
iter #8 - accuracy: 1
iter #9 - accuracy: 1
Average accuracy: 1
Wait untill weights will be serialized to net_11_valacc_1.000000.dat
Done
```

## Current Behavior
But! If I build same example with ```dnn_trainer<net_type> trainer(net,sgd(),{0,1})``` and run it on the same machine and same dataset I get 
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.78       Driver Version: 410.78       CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 2070    Off  | 00000000:09:00.0 Off |                  N/A |
| 26%   51C    P2   157W / 175W |   1037MiB /  7949MiB |     91%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce RTX 2070    Off  | 00000000:0A:00.0 Off |                  N/A |
| 26%   51C    P2   153W / 175W |   1035MiB /  7952MiB |     93%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     19817      C   Learner                                     1025MiB |
|    1     19817      C   Learner                                     1023MiB |
+-----------------------------------------------------------------------------+
```
I  got:
```
trainobjs.size(): 2
validobjs.size(): 2
step#: 0     learning rate: 0.1   average loss: 0            steps without apparent progress: 0
step#: 1400  learning rate: 0.1   train loss: 0.692944     test loss: 0.69333      steps without apparent progress: train=511, test=123
step#: 2898  learning rate: 0.1   train loss: 0.690208     test loss: 0.691613     steps without apparent progress: train=134, test=158
step#: 4385  learning rate: 0.01  train loss: 0.690187     test loss: 0.691566     steps without apparent progress: train=1851, test=658
step#: 5875  learning rate: 1e-05  train loss: 0.693145     test loss: 0.692599     steps without apparent progress: train=459, test=913
Training has been accomplished
Saved state to ./trainer_10_sync
Accuracy evaluation on validation set:
iter #0 - accuracy: 0.5
iter #1 - accuracy: 0.5
iter #2 - accuracy: 0.5
iter #3 - accuracy: 0.5
iter #4 - accuracy: 0.5
iter #5 - accuracy: 0.5
iter #6 - accuracy: 0.5
iter #7 - accuracy: 0.5
iter #8 - accuracy: 0.5
iter #9 - accuracy: 0.5
Average accuracy: 0.5
Wait untill weights will be serialized to net_10_valacc_0.500000.dat
Done
```
What may cause this effect?

## Steps to Reproduce
build **dnn_introduction2_ex** with:
```
dnn_trainer<net_type,adam> trainer(net,adam(0.0005, 0.9, 0.999), {0,1})
```
and run on machine with two GPU

* **Version**: master 84b7227
* **Where did you get dlib**: here, on github
* **Platform**: ubuntu server 18.04x64
* **Compiler**: gcc version 7.3.0

**Comment by davisking** [comment]
Dnn training can be unstable.  Splitting across GPUs effects how batch norm is computed, and maybe that's enough to make your network unstable.  I can't say.  Maybe you need a smaller learning rate.  But I don't know.

**Comment by pi-null-mezon** [comment]
Does it mean, that if I will remove all batch normalization layers from the network, learning on two GPUs can stabilize?

**Comment by davisking** [comment]
Batch norm is good to have. I’m just saying it’s the only thing that should be any different when using multiple gpus.
